conans/model/build_info.py

```diff
@@ -107,32 +227,143 @@ class CppInfo(_CppInfo):
 
+    def _check_dirs_values(self):
+        msg_template = "Using Components and global '{}' values ('{}') is not supported"
+        for dir_name in self._default_dirs_values:
+            dirs_value = getattr(self, dir_name)
+            if dirs_value is not None and dirs_value != self._default_dirs_values[dir_name]:
+                raise ConanException(msg_template.format(dir_name, dirs_value))
+
+    def _clear_dirs_values(self):
+        for dir_name in self._default_dirs_values:
+            if getattr(self, dir_name) == self._default_dirs_values[dir_name]:
+                self.__dict__[dir_name] = None
+
+    def __getitem__(self, key):
+        if self._libs or self._exes:
+            raise ConanException("Usage of Components with '.libs' or '.exes' values is not allowed")
+        self._clear_dirs_values()
+        self._check_dirs_values()
+        if key not in self._components:
+            self._components[key] = Component(key, self.rootpath)
+        return self._components[key]
+
+    @property
+    def components(self):
+        return self._components
+
     def __getattr__(self, config):
 
         def _get_cpp_info():
             result = _CppInfo()
             result.rootpath = self.rootpath
             result.sysroot = self.sysroot
-            result.includedirs.append(DEFAULT_INCLUDE)
-            result.libdirs.append(DEFAULT_LIB)
-            result.bindirs.append(DEFAULT_BIN)
-            result.resdirs.append(DEFAULT_RES)
-            result.builddirs.append("")
+            result.includedirs.extend(self._default_dirs_values["includedirs"])
+            result.libdirs.extend(self._default_dirs_values["libdirs"])
+            result.bindirs.extend(self._default_dirs_values["bindirs"])
+            result.resdirs.extend(self._default_dirs_values["resdirs"])
+            result.builddirs.extend(self._default_dirs_values["builddirs"])
             return result
 
         return self.configs.setdefault(config, _get_cpp_info())
 
 
+class Component(object):
+
+    def __init__(self, name, root_folder):
+        self._rootpath = root_folder
+        self.name = name
+        self.deps = []
+        self._lib = None
+        self._exe = None
+        self.system_deps = []
+        self.includedirs = [DEFAULT_INCLUDE]
+        self.libdirs = [DEFAULT_LIB]
+        self.resdirs = [DEFAULT_RES]
+        self.bindirs = [DEFAULT_BIN]
+        self.builddirs = [DEFAULT_BUILD]
+        self.srcdirs = []
+        self.defines = []
+        self.cflags = []
+        self.cppflags = []
+        self.cxxflags = []
+        self.sharedlinkflags = []
+        self.exelinkflags = []
+        self._filter_empty = True
+
+    def _filter_paths(self, paths):
+        abs_paths = [os.path.join(self._rootpath, p) for p in paths]
+        if self._filter_empty:
+            return [p for p in abs_paths if os.path.isdir(p)]
+        else:
+            return abs_paths
+
+    @property
+    def lib(self):
+        return self._lib
```

Comment:
are we allowing only 1 lib/exe per component?

Reply:
Yes, if still with a component you have more than one... come on, inception movie.